---
title: "Analyze data using ArchR"
author: "Zhijian Li"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(stringr)
library(magrittr)
library(WriteXLS)
library(tidyr)
library(dplyr)
library(plotly)
library(cluster)
library(cowplot)
library(gridExtra)
library(viridis)
library(GenomicRanges)
library(GenomeInfoDb)
library(data.table)
library(ArchR)
```

```{r set_parameters, echo=FALSE}
## set parameters
set.seed(42)
addArchRThreads(threads = 1)
addArchRGenome("hg38")


cols.celltypes <- c(
    'Fibroblasts 0' = '#f032e6',
    'Fibroblasts 1' = '#911eb4',
    'Fibroblasts 2' = '#e6beff',
    'Fibroblasts 3' = '#f032e6',
    'Fibroblasts 5' = '#f032e6')
```

## Reading data
```{r}
proj <- loadArchRProject(path = "./Fib", showLogo = FALSE)
```


## Identifying Marker Genes without batch correction
```{r}
markersGS <- getMarkerFeatures(
    ArchRProj = proj, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    testMethod = "wilcoxon"
)

saveRDS(markersGS, file = "./Fib/Marker_Genes.Rds")

markerList <- getMarkers(markersGS, cutOff = "FDR <= 0.01 & Log2FC >= 0.5")
markerList <- lapply(markerList, as.data.frame)

for(i in 1:length(markerList)){
    markerList[[i]] <- markerList[[i]][order(-markerList[[i]]$Log2FC), ]
}

WriteXLS::WriteXLS(markerList,
                   ExcelFileName = "./Fib/Marker_Genes.xlsx",
                   SheetNames = names(markerList))

heatmapGS <- plotMarkerHeatmap(
  seMarker = markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 0.5", 
  transpose = TRUE,
  clusterCols = TRUE,
  limits = c(-1.5, 1.5))

ComplexHeatmap::draw(heatmapGS, 
                     heatmap_legend_side = "bot", 
                     annotation_legend_side = "bot")
```

## Identifying Marker Genes after batch correction
```{r}
markersGS <- getMarkerFeatures(
    ArchRProj = proj, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters_Harmony",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    testMethod = "wilcoxon",
    maxCells = 10000
)

markerList <- getMarkers(markersGS, 
                         cutOff = "FDR <= 0.01 & Log2FC >= 0.5")
markerList <- lapply(markerList, as.data.frame)

saveRDS(markersGS, file = "./Fib/Marker_Genes_Harmony.Rds")

for(i in 1:length(markerList)){
    markerList[[i]] <- markerList[[i]][order(-markerList[[i]]$Log2FC), ]
}

WriteXLS::WriteXLS(markerList,
                   ExcelFileName = "./Fib/Marker_Genes_Harmony.xlsx",
                   SheetNames = names(markerList))

heatmapGS <- plotMarkerHeatmap(
  seMarker = markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 0.5", 
  transpose = TRUE,
  clusterCols = TRUE,
  limits = c(-1.5, 1.5))

ComplexHeatmap::draw(heatmapGS, 
                     heatmap_legend_side = "bot", 
                     annotation_legend_side = "bot")
```


## add peaks and motifs
```{r}
message("detect peaks add motif analysis")

pathToMacs2 <- findMacs2()

proj <- addGroupCoverages(ArchRProj = proj, 
                          groupBy = "Clusters_Harmony",
                          force = TRUE)

proj <- addReproduciblePeakSet(
    ArchRProj = proj, 
    pathToMacs2 = pathToMacs2,
    groupBy = "Clusters_Harmony",
    force = TRUE
)

proj <- addPeakMatrix(proj)

getAvailableMatrices(proj)

proj <- addBgdPeaks(proj)

```

## add marker peaks
```{r}
markersPeaks <- getMarkerFeatures(
    ArchRProj = proj, 
    useMatrix = "PeakMatrix", 
    groupBy = "Clusters_Harmony",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

saveRDS(markersPeaks, file = "./Fib/Marker_Peaks_Harmony.Rds")

heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markersPeaks, 
  cutOff = "FDR <= 0.05 & Log2FC >= 1", 
  transpose = TRUE,
  clusterCols = TRUE,
  limits = c(-1.5, 1.5))

ComplexHeatmap::draw(heatmapPeaks, 
     heatmap_legend_side = "bot", 
     annotation_legend_side = "bot")
```


## save data
```{r}
saveArchRProject(ArchRProj = proj, 
                 load = FALSE)
```

## Session information
```{r}
sessionInfo()
```
