---
title: "Cell Clustering"
author: "Zhijian Li"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: html_document
params:
    sample: "SAMPLE"
---

```{r setup, include=FALSE}
library(ggplot2)
library(stringr)
library(magrittr)
library(WriteXLS)
library(tidyr)
library(dplyr)
library(plotly)
library(Signac)
library(Seurat)
library(cluster)
library(clustree)
library(mclust)
library(cowplot)
library(gridExtra)
library(ggrastr)
library(viridis)
library(GenomicRanges)
library(GenomeInfoDb)
library(BSgenome.Hsapiens.UCSC.hg38)
library(EnsDb.Hsapiens.v86)
library(data.table)
library(patchwork)
library(foreach)
library(doParallel)
```

```{r set_parameters, echo=FALSE}
sample <- params$sample

atac_to_rna <- setNames(c("CK158", "CK159", "CK160",  "CK162", 
                         "CK163", "CK164","CK165"), 
                       c("CK166", "CK167", "CK168",  "CK170", 
                         "CK171", "CK173", "CK174"))

atac_dir <- paste0("../../run20_ATAC_visium/", sample) 
rna_dir <- "../../scRNA_filtered"

dir.out <- paste0("../data/", sample)
if(!dir.exists(dir.out)){
    dir.create(dir.out)
}

rna_sample <- atac_to_rna[sample]
print(paste("SAMPLE, atacname, rna_name:", sample, rna_sample))
```


## define color code
```{r}
cols = c('Cardiomyocytes' = '#800000',
'Cardiomyocytes 1' = '#800000',
'Cardiomyocytes 2' = '#9A6324',
'Cardiomyocytes 3' = '#808000',
'Fibroblasts' = '#911eb4',
'Fibroblasts 1 COL15A1+' = '#911eb4',
'Fibroblasts 1' = '#911eb4',
'Fibroblasts 2 SCARA5+' = '#e6beff',
'Fibroblasts 2' = '#e6beff',
'Fibroblasts 3' = '#f032e6',
'Endothelial cells' = '#000075',
'Endothelial cells 1' = '#000075',
'Endothelial cells 2' = 'blue',
'Endothelial cells 2 POSTN+' = 'blue',
'Endothelial cells 3' = '#568198',
'Endothelial cells 3 PLVAP+' = '#568198',
'Endothelial cells 3 VEGFC+' = '#568198',
'Endothelial cells 4' = '#469990',
'Endothelial cells 4 SEMA3G+' = '#469990',
'Macrophages' = '#e6194B',
'Macrophages 1 CD163+' = '#e6194B',
'Macrophages 2 CD11C+' = '#fabebe',
'Pericytes' = '#f58231',
'Pericytes EGFLAM+' = '#f58231',
'T cells' = '#ffe119',
'Lymphatic endothelial cells' = '#ffd8b1',
'Adipocytes' = '#000000',
'Neuronal cells' = '#42d4f4',
'Erythrocytes' = '#999999',
'Proliferating cells' = '#999999',
'Damaged endothelial cells' = '#999999',
'Vascular smooth muscle cells' = '#aaffc3')
```


## define celltype name
```{r}
cell.names = c(
    'Cardiomyocytes' = 'Cardiomyocytes',
    'Cardiomyocytes 1' = 'Cardiomyocytes',
    'Cardiomyocytes 2' = 'Cardiomyocytes',
    'Cardiomyocytes 3' = 'Cardiomyocytes',
    'Fibroblasts' = 'Fibroblasts',
    'Fibroblasts 1 COL15A1+' = 'Fibroblasts_1',
    'Fibroblasts 1' = 'Fibroblasts_1',
    'Fibroblasts 2 SCARA5+' = 'Fibroblasts_2',
    'Fibroblasts 2' = 'Fibroblasts_2',
    'Fibroblasts 3' = 'Fibroblasts_3',
    'Endothelial cells' = 'Endothelial',
    'Endothelial cells 1' = 'Endothelial_1',
    'Endothelial cells 2' = 'Endothelial_2',
    'Endothelial cells 2 POSTN+' = 'Endothelial_2',
    'Endothelial cells 3' = 'Endothelial_3',
    'Endothelial cells 3 PLVAP+' = 'Endothelial_3',
    'Endothelial cells 3 VEGFC+' = 'Endothelial_3',
    'Endothelial cells 4' = 'Endothelial_4',
    'Endothelial cells 4 SEMA3G+' = 'Endothelial_4',
    'Macrophages' = 'Macrophages',
    'Macrophages 1 CD163+' = 'Macrophages_1',
    'Macrophages 2 CD11C+' = 'Macrophages_2',
    'Pericytes' = 'Pericytes',
    'Pericytes EGFLAM+' = 'Pericytes_EGFLAM',
    'T cells' = 'T_cells',
    'Lymphatic endothelial cells' = 'Lymphatic_Endo',
    'Adipocytes' = 'Adipocytes',
    'Neuronal cells' = 'Neuronal',
    'Erythrocytes' = 'Erythrocytes',
    'Proliferating cells' = 'Pro_cells',
    'Damaged endothelial cells' = 'Damaged_Endo',
    'Vascular smooth muscle cells' = 'VSMCs')
```


## processing ATAC-seq data
```{r load_data}
heart.atac <- readRDS(paste0(dir.out, "/", sample, ".Rds"))
fragment.path <- paste0(atac_dir, "/outs/fragments.tsv.gz")
```

## peak calling for each predicted label
```{r}
Idents(heart.atac) <- "predicted.id"
heart.atac <- RenameIdents(heart.atac, cell.names)
heart.atac$annotation <- Idents(heart.atac)

print(unique(heart.atac$annotation))

fragment_dir <- paste0(dir.out, "/Fragments")
peaks_dir <- paste0(dir.out, "/Peaks")
if(!dir.exists(fragment_dir)){
    dir.create(fragment_dir)
}

if(!dir.exists(peaks_dir)){
    dir.create(peaks_dir)
}

for(anno in unique(heart.atac$annotation)) {
    fragment_file_filtered <- paste0(fragment_dir, "/", anno, ".tsv")
    if(file.exists(fragment_file_filtered)){
        file.remove(fragment_file_filtered)
    }
    
    cells <- colnames(subset(heart.atac, subset = annotation == anno))
    FilterFragments(fragment.path = fragment.path,
                    cells = cells,
                    output.path = fragment_file_filtered)
    
    command <- sprintf("macs2 callpeak -g hs --name %s --treatment %s --outdir %s --format BED --nomodel --call-summits --nolambda --keep-dup all", 
	    anno, 
	    paste0(fragment_file_filtered, ".bgz"), 
	    peaks_dir)
    message("Running Macs2...")
	message(command)
	system(command, intern = TRUE)
}
```


## define helper function
```{r}
readSummits <- function(file){
    df <- data.frame(readr::read_tsv(file,
                                     col_names = c("chr","start","end","name","score")))
    df <- df[,c(1,2,3,5)] #do not keep name column it can make the size really large
    return(makeGRangesFromDataFrame(df = df,
                                    keep.extra.columns = TRUE,
                                    starts.in.df.are.0based = TRUE))
}

clusterGRanges <- function(gr, filter = TRUE, by = "score", decreasing = TRUE){
      gr <- sort(sortSeqlevels(gr))
      r <- GenomicRanges::reduce(gr, min.gapwidth=0L, ignore.strand=TRUE)
      o <- findOverlaps(gr,r)
      mcols(gr)$cluster <- subjectHits(o)
      gr <- gr[order(mcols(gr)[,by], decreasing = decreasing),]
      gr <- gr[!duplicated(mcols(gr)$cluster),]
      gr <- sort(sortSeqlevels(gr))
      mcols(gr)$cluster <- NULL
      return(gr)
}

nonOverlappingGRanges <- function(gr, by = "score", decreasing = TRUE, verbose = FALSE){
    stopifnot(by %in% colnames(mcols(gr)))
    i <-  0
    gr_converge <- gr
    while(length(gr_converge) > 0){
      if(verbose){
        message(".", appendLF = FALSE)
      }
      i <-  i + 1
      gr_selected <- clusterGRanges(gr = gr_converge, filter = TRUE, by = by, decreasing = decreasing)
      gr_converge <- subsetByOverlaps(gr_converge ,gr_selected, invert=TRUE) #blacklist selected gr
      if(i == 1){ #if i=1 then set gr_all to clustered
        gr_all <- gr_selected
      }else{
        gr_all <- c(gr_all, gr_selected)
      } 
    }
    if(verbose){
      message("\nSelected ", length(gr_all), " from ", length(gr))
    }
    gr_all <- sort(sortSeqlevels(gr_all))
    return(gr_all)

}
```


## Make Non-Overlapping Peak Set
```{r}
blacklist_file = "../../../Blacklist/lists/hg38-blacklist.v2.bed.gz"
blacklist <- rtracklayer::import.bed(blacklist_file)

chromSizes <- GRanges(names(seqlengths(BSgenome.Hsapiens.UCSC.hg38)), 
                      IRanges(1, seqlengths(BSgenome.Hsapiens.UCSC.hg38)))
chromSizes <- GenomeInfoDb::keepStandardChromosomes(chromSizes,
                                                    pruning.mode = "coarse")
      

peaks_files <- list.files(peaks_dir, 
                          pattern = "\\_summits.bed", 
                          full.names = TRUE)

gr_list <- GenomicRangesList(lapply(peaks_files, function(x){
        extended_summits <- readSummits(x) %>%
            resize(., width = 2 * 250 + 1, fix = "center") %>%
            subsetByOverlaps(.,chromSizes,type="within") %>%
              subsetByOverlaps(.,blacklist,invert=TRUE) %>%
              nonOverlappingGRanges(., by="score", decreasing=TRUE)
        extended_summits <- extended_summits[order(extended_summits$score, 
                                                   decreasing=TRUE)]
        extended_summits <- head(extended_summits, 100000)
        mcols(extended_summits)$scoreQuantile <-trunc(rank(mcols(extended_summits)$score)) / length(mcols(extended_summits)$score)
        extended_summits
        }))


unionPeaks <- nonOverlappingGRanges(unlist(gr_list), 
                                    by = "scoreQuantile", 
                                    decreasing = TRUE)
unionPeaks <- sort(sortSeqlevels(unionPeaks))

unionPeaks <- unionPeaks[seqnames(unionPeaks) %in% paste0("chr",c(1:22,"X"))]
unionPeaks <- keepSeqlevels(unionPeaks, paste0("chr",c(1:22,"X")))

df <- data.frame(seqnames=seqnames(unionPeaks),
                 starts=start(unionPeaks)-1,
                 ends=end(unionPeaks))

write.table(df, file = paste0(dir.out, "/unionPeaks.bed"), 
            sep = "\t", row.names = FALSE,
            col.names = FALSE, quote = FALSE)
```


```{r}
countInsertions <- function(query, fragments, by = "RG"){
    #Count By Fragments Insertions
    inserts <- c(
        GRanges(seqnames = seqnames(fragments), ranges = IRanges(start(fragments), start(fragments)), RG = mcols(fragments)[,by]),
        GRanges(seqnames = seqnames(fragments), ranges = IRanges(end(fragments), end(fragments)), RG = mcols(fragments)[,by])
    )
    by <- "RG"
    overlapDF <- DataFrame(findOverlaps(query, inserts, ignore.strand = TRUE, maxgap=-1L, minoverlap=0L, type = "any"))
    overlapDF$name <- mcols(inserts)[overlapDF[, 2], by]
    overlapTDF <- transform(overlapDF, id = match(name, unique(name)))
    #Calculate Overlap Stats
    inPeaks <- table(overlapDF$name)
    total <- table(mcols(inserts)[, by])
    total <- total[names(inPeaks)]
    frip <- inPeaks / total
    #Summarize
    sparseM <- Matrix::sparseMatrix(
        i = overlapTDF[, 1], 
        j = overlapTDF[, 4],
        x = rep(1, nrow(overlapTDF)), 
        dims = c(length(query), length(unique(overlapDF$name))))
    colnames(sparseM) <- unique(overlapDF$name)
    return(sparseM)
}
```



## create new count matrix
```{r}
message("Reading in fragment files...")
fragments <- data.frame(readr::read_tsv(fragment.path, col_names=FALSE))
fragments <- GRanges(
    seqnames = fragments[,1], 
    IRanges(fragments[,2]+1, fragments[,3]), 
    RG = fragments[,4], 
    N = fragments[,5]
)

fragments <- fragments[fragments$RG %in% colnames(heart.atac)]

#Create Counts matirx
counts <- countInsertions(unionPeaks, fragments, by = "RG")
rownames(counts) <- paste(seqnames(unionPeaks),
                          start(unionPeaks),
                          end(unionPeaks),
                          sep="_")
counts <- counts[rowSums(counts) > 0, ]
```


## save data
```{r}
saveRDS(counts, file = paste0(dir.out, "/unionPeaks_matrix.Rds"))
```

## Session information
```{r}
sessionInfo()
```
